tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/vmt/EUFRule.scala

```diff
@@ -0,0 +1,148 @@
+
+  /**
+   * Given a function f: (S1,...,Sn) => S, with the rule f(y1,...,yn) = ef, args `x1: S1, ..., xn: Sn`, and an
+   * expression eg, constructs a function definition for g, with the rule:
+   * ```
+   * g(y1, ... yn) = if y1 = x1 /\ ... /\ yn = xn
+   *                 then eg
+   *                 else ef
+   * ```
+   * @param fnArgTerms
+   *   the values `x1, ..., xn`
+   * @param newCaseTerm
+   *   the term `eg`
+   * @param args
+   *   pairs `(y1, S1),...,(yn,Sn)`
+   * @param baseCaseTerm
+   *   the term `ef`
+   * @return
+   */
+  private def exceptAsNewFunDef(
+      fnArgTerms: List[Term],
+      newCaseTerm: Term,
+    )(args: List[(String, Sort)],
+      baseCaseTerm: Term): FunDef = {
+    // sanity check
+    assert(args.length == fnArgTerms.length)
+
+    // Except redefines at a point <<x1, ..., xn>>. We decompose <<y1,...,yn>> = <<x,...xn>>
+    // into y1 = x1 /\ ... /\ yn = xn
+    val matchConds = args.zip(fnArgTerms).map { case ((name, sort), term) =>
+      Equal(mkVariable(name, sort), term)
+    }
+
+    // We could use a generic And-wrapper, but it's more convenient to handle nullary and unary cases separately
+    val ifCondition = matchConds match {
+      case Nil      => True // Should never happen, included for case-completeness
+      case h :: Nil => h
+      case _        => And(matchConds: _*)
+    }
+
+    FunDef(gen.newName(), args, ITE(ifCondition, newCaseTerm, baseCaseTerm))
+  }
+
+  // Convenience shorthand
+  private def rewrite: TlaEx => Term = rewriter.rewrite
+
+  // Assume isApplicable
+  override def apply(ex: TlaEx): Term =
+    ex match {
+      case OperEx(TlaOper.eq | ApalacheOper.assign, lhs, rhs) =>
+        // := is just = in VMT
+        Equal(rewrite(lhs), rewrite(rhs))
+
+      case OperEx(TlaFunOper.funDef, e, varsAndSets @ _*)
+          // All domain-defining sets must be restricted.
+          if TlaOper.deinterleave(varsAndSets)._2.forall(isRestrictedSet) =>
+        val (vars, sets) = TlaOper.deinterleave(varsAndSets)
+        val setSorts = sets.map(restrictedSetJudgement.getSort)
+        // Construct pairs of formal parameter names and sorts
+        val argList = vars.zip(setSorts).toList.map { case (NameEx(name), sort) =>
+          (name, sort)
+        }
+        FunDef(gen.newName(), argList, rewrite(e))
+      case OperEx(TlaFunOper.app, fn, arg) =>
+        val fnTerm = rewrite(fn)
+        // Arity 2+ functions pack their arguments as a single tuple, which we might need to unpack
+        arg match {
+          case OperEx(TlaFunOper.tuple, args @ _*) => Apply(fnTerm, args.map(rewrite): _*)
+          case _                                   => Apply(fnTerm, rewrite(arg))
+        }
+
+      case OperEx(TlaFunOper.except, fn, arg, newVal) =>
+        val valTerm = rewrite(newVal)
+        // Toplevel, arg is always a TLaFunOper.tuple. within, it's either a single value, or another
```

Comment:
I may be reading this wrong, but it seems like your comment isn't disagreeing with what's written in the code. Or did you meant to say that it's always a flat tuple? Because that's not true.

Reply:
I think it is always a tuple in `EXCEPT`, even in the case of a single-index expression. See: https://github.com/informalsystems/apalache/blob/e20a107b7c2766adde168148d6407dd8b64c3877/tla-io/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala#L1091-L1159

Reply:
https://github.com/informalsystems/apalache/blob/e20a107b7c2766adde168148d6407dd8b64c3877/tla-io/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala#L1139

Reply:
Yes, single index becomes `tuple` and multi-index becomes `tuple(tuple(...))`. Which is what the comment in the code says.

Reply:
I don't understand what you're trying to say.

Reply:
Oh, this `tuple(tuple(...))` structure is weird. I thought that it would be only one tuple in E3.